### ISSUE TITLE
Make `sh:prefixes` optional in SPARQL queries

### ISSUE THREAD

**Issue by HolgerKnublauch** [w3c/data-shapes#176]
Basically every user of SHACL-SPARQL ever has run into this limitation: Not only is the syntax of sh:prefixes complex to use, the expectation is that the namespace prefixes from the shapes graph apply automatically.

When this was designed, the WG decided to not rely on automatic definition of prefixes because prefixes are not persisted in the RDF Graph data model, but are just a temporary feature in during parsing. This is despite the fact that many APIs such as Jena do persist the prefixes together with the graph object. But yeah, the problem remains that when SHACL-SPARQL shapes are moved around the prefixes may get lost.

A solution here may be to automatically prepend all prefixes from the shapes graph when no sh:prefixes triple exists. We need to come up with some creative solution on how to define what these prefixes must be. But it's a huge obstacle, so we cannot just leave the current solution in place. At least I believe we can define a list of default prefixes such as rdf: owl: and sh: that should always be present.

**Comment by VladimirAlexiev** [comment]
:+1:

Repo namespaces are not 100% "reliable" because if there's a conflicting prefix, the repo can remember only one of the namespaces.
(i.e. if you load file1.ttl that defines `foo:` and the repo remembers its namespace, then when you load file2.ttl with the same prefix but different namespace, the repo won't remember it).

But if there are such conflicts, the user can override with `sh:prefixes`

**Comment by TallTed** [comment]
I do not believe we can, nor should, define such a list. There is no reason why SHACL users should be forbidden from using `owl:` elsewhere (and perhaps `owl-ns:` for `http://www.w3.org/2002/07/owl#`).

This "problem" is not specific to SHACL, and should not be addressed by SHACL-related specs. If anywhere, this "problem" should be addressed in the context of SPARQL, as there would then need to be a "SPARQL prefixes registry" or similar.

At present, there is no IANA nor W3C "prefix registry", which would be necessary to prevent collisions. The closest thing of which I'm aware is the prefix.cc lookup service — which does not prevent collisions!


> This is despite the fact that many APIs such as Jena do persist the prefixes together with the graph object.

Those APIs are acting outside of all prefixed-name specifications of which I am aware, which specify that the prefixes be declared in the same document in which they are used. I certainly hope that Jena and any other software that acts this way treats prefix declarations they encounter as overriding their persisted prefixes in the context of the live declaration.

How does Jena handle a Turtle document that includes multiple declarations of the same prefix? Which declaration does Jena persist, and for how long? Does Jena persist the first declaration it encounters for any given prefix, over-riding later declarations in the same Turtle document? Or does Jena persist the "last declaration from the first document that contains a declaration for that prefix"?

Note that users have the option of setting *any prefix they like* in a given document, and may use the same prefix with multiple namespaces in a single Turtle document, among other places. Note that this is explicitly permitted by the Turtle spec, and each declaration is active for the prefixed names following that declaration and preceding any other declaration of the same prefix.

`rdf:` could be used just as well for `http://example.org/rough-data-format#` as for [`http://www.w3.org/1999/02/22-rdf-syntax-ns#` or `https://cacax.fun/`](https://prefix.cc/rdf). 

`sh:` could be `http://shell.example#` and/or `http://example.sh#` as well as [`http://www.w3.org/ns/shacl#` or `http://purl.org/skos-history/`](https://prefix.cc/sh).

> At least I believe we can define a list of default prefixes such as `rdf:`, `owl:` and `sh:` that should always be present.

Such a requirement would conflict with the rest of the universe of RDF specifications, even if nowhere else — but I believe it would conflict with many other specs as well.

**Comment by HolgerKnublauch** [comment]
Whatever solution we want to implement here, SOMETHING has to be done. The current syntax of SHACL with the sh:prefixes has not passed the test of time and basically every user has problems with this. This is an obstacle to wider adoption of SHACL-SPARQL. We can find many reasons not to do something, yet we should have an open discussion with pragmatics as one of the main drivers.

**Comment by VladimirAlexiev** [comment]
- There is https://w3c.github.io/json-ld-rc/ and eg https://w3c.github.io/json-ld-rc/prefixes.ttl (source at https://github.com/w3c/json-ld-rc).
  - It includes a few too many prefixes to my liking (and see https://github.com/w3c/json-ld-rc/issues/6)
  - (At least it doesn't define `geo`, masking out the `geo:` URI scheme, which I have seen elsewhere)
  - It lacks `sh` (posted https://github.com/w3c/json-ld-rc/issues/38)
- @TallTed many repos (including GraphDB) remember a prefix the first time they see it, in the repo's Namespaces. (Only the first occurrence is remembered.)
Then the SPARQL editor can auto-insert such prefix when used in a query. It's a great usability feature.
- If a user wants to redefine a system prefix, they can do so: they just need to list it in `sh:prefixes`

**Comment by HolgerKnublauch** [comment]
Yes, or they can put explicit PREFIX declarations into the sh:select string. The defaults would only serve as fallback.

**Comment by TallTed** [comment]
@VladimirAlexiev — Whether "[remembering] a prefix the first time they see it" is "a great usability feature" depends greatly on whether that first-use is optimal for the users of that deployment. Indeed, given that, for instance, Turtle mandates that the _latest_ declaration of a given prefix cover any given occurrence within the instance data, "Only the first occurrence is remembered" seems like a great usability *negative*. At the least, there should be some way to tell such an "auto-remembering" system that *this* declaration should replace a previously remembered declaration and/or to forget all previously remembered declarations.

Indeed, unless there's some point of user confirmation, "the SPARQL editor can auto-insert such prefix when used in a query" is likely to lead to confusing results that may not match results on any other SPARQL processor, including other deployments of the same software with the same *loaded* data, just because user queries were run in a different order following data load (i.e., Query_1 on Server_1 uses prefix `sh:` with one namespace, while Query_2 on Server_1 uses prefix `sh:` with a different namespace, and these two queries are run in reverse order on Server_2).

Where do you learn what prefix declarations were used for a given query?

This seems like a pure landmine to me.

fwiw, Virtuoso has a table of prefix/namespace registrations (visible [here](https://dbpedia.org/sparql/?help=nsdecl) for the DBpedia instance). These [can be set](https://docs.openlinksw.com/virtuoso/fn_xml_set_ns_decl/) for use in exports to media types that support prefixed names (such as N3 or Turtle), and to be a fallback for a SPARQL query that doesn't include one or more declarations for prefixed names found in that query. Declarations that occur in a query over-ride those in the table, when the same prefix is found in both places. There are rules in SPARQL and Turtle (among other places) that govern handling of duplicate prefixes that are declared with different namespaces within the same query or document.

I know that these stored namespaces can lead to user confusion because they *have*. This is another downside to the optimistic SPARQL interaction that lacks any useful way to report errors or communicate other things (such as predefined namespace prefixes that it applied to execution of a given query) to the user outside of HTTP headers (which many users never see). We provide the facility to predefine declarations because of user demand; I have hopes that this feature will be improved over time to decrease such user confusion.

**Comment by afs** [comment]
A partial solution would be to have a defined prefixes declaration that applies to the whole shapes graph if there is no `sh:prefixes` for the particular SPARQL usage. It could even be a string in Turtle syntax for convenience

Requiring shapes graph Turtle prefixes handling
* would exclude N-Triples
* is unclear as to what happens in the presence of imports (do they inherit?)

**Comment by HolgerKnublauch** [comment]
I am closing this after not finding a better solution, and because nobody else seems to have a better solution either.

- We could potentially define a series of default prefixes but those will be highly controversial. If I just check what Valdimir pointed out at https://github.com/w3c/json-ld-rc I would already challenge many decisions there. There is little chance to get a wider community agree on a definite list.
- We cannot rely on prefixes from serializations as they are not persisted anywhere.

So to me this becomes a tooling issue. Validation engines need to clearly indicate errors and suggest solutions. Editing tools need to automatically insert the requires prefix declarations. Training material needs to hammer the point down too.

If anyone has better ideas, please reopen.

**Comment by HolgerKnublauch** [comment]
I now think there is a better solution:

If a SHACL-SPARQL query does NOT have any sh:prefixes triple, then collect all sh:PrefixDeclarations from the current shapes graph and insert them to the beginning of the query string.

This means that something like

```
ex:PrefixDeclaration
    sh:prefix "ex" ;
    sh:namespace "http://example.com/ns#"^^xsd:anyURI ;
.

ex:SomeShape
    sh:sparql [
        sh:select "SELECT $this WHERE { $this a ex:Class }" ;
    ] .
```

would be valid.

And I don't see how this could break backwards-compability. If a query string already declares PREFIXes on top, they will take precedence over the additional ones that the engine will insert.

**Comment by afs** [comment]
The problem statement has two parts -

> Not only is the syntax of sh:prefixes complex to use, the expectation is that the namespace prefixes from the shapes graph apply automatically.

The second part, shapes graph prefixes, is hard to standardize as has already been discussed.

For the first part, the prefix declarations, maybe this can be made easier. Some possibilities:

Provide a micro-syntax:
* `sh:declare "PREFIX ex: <http://example/>" ;`
* ```
  sh:declare """
     PREFIX ex:     <http://example/>
     PREFIX schema <https://schema.org/>
     """ ;
   ```

Convenience quoted URIs:
* Drop the requirement for `^^xsd:anyURI` - allow a plain string as well (which has shacl-shacl checking changes).

Abbreviated syntax:
* `sh:declare ("ex", "http://example/")`
* `sh:declare ("ex", "http://example/"), ("schema" "https://schema.org/") ;`

The example in SHACL-sparql becomes:
```turtle
ex:dftPrefixes
    sh:declare ( "ex" "http://example.com/ns#" ) ;
    sh:declare ( "schema"  "https://schema.org/" ) ;
    ] ;
.
```

Allow inline declarations:

```turtle
ex:SomeShape
    sh:declare ( "foaf", "http://xmlns.com/foaf/0.1/") ;
    sh:sparql [
        sh:select "SELECT $this WHERE { $this foaf:name 'Alice' }" ;
    ] .
```

> If a SHACL-SPARQL query does NOT have any sh:prefixes triple,

you can have a global (shared) and add local additions:

```turtle
ex:SomeShape
    sh:prefixes ex:dftPrefixes ;
    sh:declare ( "foaf", "http://xmlns.com/foaf/0.1/") ;
    sh:sparql [
        sh:select "SELECT $this WHERE { $this foaf:name 'Alice' }" ;
    ] .
```

**Comment by HolgerKnublauch** [comment]
@afs thanks for the suggestions. I definitely agree with allowing xsd:string at sh:namespace. The recent PR includes that change alongside the handling of the case when no sh:prefixes are declared.

On the syntactic sugar, I have mixed feelings. I don't think the main problem was that people found the syntax too hard to enter these, but rather that people were simply not aware that they need to declare them, and then the handling of owl:imports is quite advanced.

Having too many syntactic alternatives can make processing harder. In particular, right now an engine can simply scan the graph for any triple with predicate sh:prefix and find all declarations. This would be far more complex for all the other suggestions.

(And your last example with sh:declare ( "foaf" etc) has too little benefits compared to simply declaring it in the sh:select string itself.)

**Comment by bergos** [comment]
The proposal collects all prefix declarations scattered across the shapes graph, which creates implicit dependencies. Consider a shapes graph with a person shape that owns its own prefix declarations:

```turtle
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix ex: <http://example.com/ns#> .

ex:PersonNS
    sh:declare [
        sh:prefix "ex" ;
        sh:namespace "http://example.com/ns#"^^xsd:anyURI ;
    ] ;
    sh:declare [
        sh:prefix "schema" ;
        sh:namespace "http://schema.org/"^^xsd:anyURI ;
    ] .

ex:PersonShape
    a sh:NodeShape ;
    sh:targetClass ex:Person ;
    sh:prefixes ex:PersonNS;
    sh:sparql [
        sh:select """SELECT $this WHERE { FILTER NOT EXISTS { $this schema:name ?name } }""" ;
        sh:message "Person must have a schema:name." ;
    ] .
```

`ex:DocumentShape` has no `sh:prefixes` triple and uses `ex:title`, which only resolves because `ex:PersonNS` happens to declare `"ex"`:

```turtle
ex:DocumentShape
    a sh:NodeShape ;
    sh:targetClass ex:Document ;
    sh:sparql [
        sh:select """SELECT $this WHERE { FILTER NOT EXISTS { $this ex:title ?title } }""" ;
        sh:message "Document must have an ex:title." ;
    ] .
```

When `ex:PersonShape` and `ex:PersonNS` are deleted, `ex:DocumentShape` loses the `"ex"` prefix declaration without any indication in its own triples. The processor prepends no `PREFIX ex:` declaration, so `ex:title` cannot be resolved → ill-formed query.

If `ex:AnotherShape` adds its own declaration for `"ex"` pointing to a different namespace, the shapes graph now contains two conflicting declarations for `"ex"`, making it ill-formed for every query without `sh:prefixes`.

```turtle
ex:AnotherNS
    sh:declare [
        sh:prefix "ex" ;
        sh:namespace "http://other.example.com/ns#"^^xsd:anyURI ;
    ] .
```

I would propose a two-level hierarchy instead:

- **Shapes graph declarations** — attached to the named graph IRI typed as `sh:ShapesGraph`, shared across all SPARQL queries in that graph.
- **Local declarations** — attached to an individual constraint via `sh:prefixes`, extending or overwriting the shapes graph declarations for that constraint only.

If a query has no `sh:prefixes` triple, the processor uses only the shapes graph declarations. Prefix declarations stay in one well-defined place by default, and individual constraints can still opt in to additional or overriding mappings when needed:

```turtle
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix ex: <http://example.com/ns#> .

<http://example.com/shapes>
    a sh:ShapesGraph ;
    sh:declare [
        sh:prefix "ex" ;
        sh:namespace "http://example.com/ns#"^^xsd:anyURI ;
    ] ;
    sh:declare [
        sh:prefix "schema" ;
        sh:namespace "http://schema.org/"^^xsd:anyURI ;
    ] .

ex:PersonShape
    a sh:NodeShape ;
    sh:targetClass ex:Person ;
    sh:sparql [
        sh:select """SELECT $this WHERE { FILTER NOT EXISTS { $this schema:name ?name } }""" ;
        sh:message "Person must have a schema:name." ;
    ] .

ex:DocumentShape
    a sh:NodeShape ;
    sh:targetClass ex:Document ;
    sh:sparql [
        sh:select """SELECT $this WHERE { FILTER NOT EXISTS { $this ex:title ?title } }""" ;
        sh:message "Document must have an ex:title." ;
    ] .
```

Deleting `ex:PersonShape` has no effect on `ex:DocumentShape`. Prefix declarations live in one well-defined place and are governed by the shapes graph, not by individual shapes.

*edit 2026-06-22: added the missing `sh:prefixes ex:PersonNS;` triple.*

**Comment by HolgerKnublauch** [comment]
@bergos how would this approach identify "the" shapes graph? There can be any number of sh:ShapesGraph instances per graph, and the engine doesn't get the URI(s) as input, so it cannot determine the "main" shapes graph URI.

In your first example I think there is also something missing, because PersonShape does *not* link to PersonNS, so I struggle to understand your point.

The issue about conflicting declarations and if someone deletes a declaration already affect the current design. These are user errors.

**Comment by nicholascar** [comment]
As per the Profiling [Packaging SHACL section](https://w3c.github.io/data-shapes/shacl12-profiling/#packaging), You can mark a Shape as being part of a particular ShapesGraph with `:ShapeX rdfs:isDefinedBy :ShapesGraphY` or `:ShapesGraphY rdfs:member :ShapeX` with the former winning in any conflicts with the latter. 

Is this helpful?

**Comment by HolgerKnublauch** [comment]
@nicholascar thanks for thinking. rdfs:isDefinedBy would probably just shift the problem as it is not present in many/most graphs that I have seen.

**Comment by nicholascar** [comment]
> rdfs:isDefinedBy would probably just shift the problem as it is not present in many/most graphs that I have seen.

Of course not as it's only now been proposed for packaging. But it does provide a mechanism to move the issue completely into user error space if wrong if selected as a solution.

**Comment by bergos** [comment]
@HolgerKnublauch If there are multiple instances of `sh:ShapesGraph`, their prefixes can be combined. However, attaching prefix declarations to an instance of `sh:ShapesGraph` makes it more explicit that the declaration is intended to be reused.

This means that if multiple `sh:ShapesGraph` instances exist in the graph, the issues I described can still arise — but since the declarations are less hidden, the likelihood of encountering these problems is significantly reduced.

**Comment by HolgerKnublauch** [comment]
Ok, I can support this. So the requirement would be to have

```
<...>
    a sh:ShapesGraph ;    # or, theoretically: subclass of sh:ShapesGraph
    sh:declare [
        sh:prefix "ex" ;
        sh:namespace "http://example.com/ns#" ;
    ] .
```

Can we anticipate any cases where prefix declarations will be used in Data Graphs? Where would, for example, rules live (see #954). I assume rules are similar to shapes, and not declared against specific instance data?

**Comment by TallTed** [comment]
Prefix declarations are not part of graphs. They are part of serializations.

I am concerned that current descriptions of how to use `sh:prefix` promote prefix declarations far beyond where they should be, especially but not only if they are "harvested" in some way that removes them from the proximate triples that use them, and if multiple declarations of the same `prefix` result in one or more of the associated `namespace` being lost.

**Comment by HolgerKnublauch** [comment]
There are many ways in which users can shoot themselves in the foot. Removing triples without considering the consequences is one of them. But this argument applies to basically every feature.

For prefixes there is IMHO clear evidence (from user questions etc) that the old approach wasn't doing a good enough job, so this ticket tried to promote a practice in which named graphs that introduce a new prefix also sh:declare it (at the sh:ShapesGraph instance). There will be a learning curve for this practice to take hold, but the good thing is that this can be automated from the prefix declarations in files, and only need to be done once per file.

(Did you mean to re-open this ticket? It's closed now with the recent PR, and I don't yet see reasons to reopen it)

**Comment by HolgerKnublauch** [comment]
I want to reopen this to generalize the mechanism from sh:ShapesGraph/RulesGraph to any known subclass* of owl:Ontology. This is because I noticed that people already have plenty of files that are currently only declared as owl:Ontology yet hold prefix declarations. In particular, our own TTL files such as shacl.ttl and shnex.ttl declare their graph URI as owl:Ontology only, not sh:ShapesGraph. So their own declarations are invisible to this mechanism.

There is no need to limit this to sh:ShapesGraph - as we learned quickly there was sh:RulesGraph introduced and the policy already need to be expanded.

**Comment by TallTed** [comment]
> There is no need to limit this to `sh:ShapesGraph`

Why is it being limited to "any known `subclass*` of `owl:Ontology`", and not open to "any known `subclass*` of `rdfs:Graph`", or an even larger superclass?

**Comment by HolgerKnublauch** [comment]
There is no such class "rdfs:Graph". owl:Ontology is rdfs:subClassOf rdfs:Resource and that superclass is too general.
